aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,376 @@
+/// This module provides a veiled coin type, denoted `VeiledCoin<T>` that hides the value/denomination of a coin.
+///
+/// Our implementation relies on a secondary so-called "resource account" which helps us mint a `VeiledCoin<T>` from a
+/// traditional `coin::Coin<T>` by transferring this latter coin into a `coin::CoinStore<T>` resource stored in the
+/// resource account. Later on, when someone wants to convert their `VeiledCoin<T>` into a traditional `coin::Coin<T>`
+/// the resource account can be used to transfer out said `coin::Coin<T>` from its coin store. This is where the
+/// "resource account" becomes important, since transfering out a coin like this requires a `signer` for the resource
+/// account, which this module can obtain via a `SignerCapability`.
+module veiled_coin::veiled_coin {
+    use std::error;
+    use std::features;
+    use std::signer;
+    use std::vector;
+    use aptos_std::pedersen::{Self, Commitment};
+    use aptos_std::ristretto255::{Self, CompressedRistretto, Scalar, new_scalar_from_u64, new_scalar_from_bytes, scalar_neg, point_equals, point_decompress};
+
+    use aptos_framework::account;
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::event::{Self, EventHandle};
+    use aptos_std::bulletproofs::RangeProof;
+    use aptos_std::bulletproofs;
+
+    //
+    // Errors.
+    //
+
+    /// The range proof system needs to support proofs for any number \in [0, 2^{64})
+    const ERANGE_PROOF_SYSTEM_HAS_INSUFFICIENT_RANGE : u64 = 1;
+
+    /// A range proof failed to verify.
+    const ERANGE_PROOF_VERIFICATION_FAILED : u64 = 2;
+
+    /// Account already has `VeiledCoinStore` registered for `CoinType`
+    const EVEILED_COIN_STORE_ALREADY_PUBLISHED: u64 = 4;
+
+    /// Account hasn't registered `CoinStore` for `CoinType`
+    const EVEILED_COIN_STORE_NOT_PUBLISHED: u64 = 5;
+
+    /// Not enough coins to complete transaction
+    const EINSUFFICIENT_BALANCE: u64 = 6;
+
+    //
+    // Constants
+    //
+
+    /// The maximum number of bits used to represent a coin's value.
+    const MAX_BITS_IN_VALUE : u64 = 64;
+
+    /// The domain separation tag (DST) used for the Bulletproofs prover.
+    const VEILED_COIN_DST : vector<u8> = b"AptosVeiledCoinExample";
+
+    //
+    // Core data structures.
+    //
+
+    /// Main structure representing a coin in an account's custody.
+    struct VeiledCoin<phantom CoinType> {
+        /// Pedersen commitment (under the default Bulletproofs CK) to a number of coins v \in [0, 2^{64}), an invariant
+        /// that is enforced throughout the code
+        private_value: Commitment,
+    }
+
+    /// A holder of a specific coin types and associated event handles.
+    /// These are kept in a single resource to ensure locality of data.
+    struct VeiledCoinStore<phantom CoinType> has key {
+        /// A Pedersen commitment to a value v \in [0, 2^{64}), an invariant that is enforced throughout the code.
+        private_balance: CompressedRistretto,
```

Comment:
Here we'll have a `public_key` field for the ElGamal PK.

And we'll add an entry function for an existing Aptos account to "register" their veiled coin store with a given PK.

Reply:
A public key field could be added to `register` for initial registration, with an additional entry function added for key rotation?